### DISCUSSION TITLE
watching HAML files seems to work adding classes... but not removing them?

### DISCUSSION THREAD

**Discussion by jf** [tailwindlabs/tailwindcss#13597] [Help]
This is a strange one to me. I have `npx tailwindcss -i src/input.css -o src/output.css --watch` going, and with the following `tailwind.config.js`:
```
/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [ 'views/**/*.haml' ],
  theme: {
    extend: {},
  },
  plugins: [],
}
```

When I *add* classes to my HAML layout file, they show up in the output CSS... but when I *remove* the *same said classes*, they *do not disappear* from the output CSS. That is, unless and until I restart `npx tailwindcss ... --watch`. I'm guessing this isn't expected behaviour?

**ALSO:** how do I find the regular expressions that tailwindcss is using? I am not entirely sure how much support there is for HAML. I was hopeful, given that I was able to have the added classes show up... but the reverse not happening is puzzling. I'm hoping to be able to find the actual regular expressions to be able to confirm the support (hopefully) for myself, but I am not able to see anything in Tailwind's [default configuration](https://github.com/tailwindlabs/tailwindcss/blob/master/stubs/config.full.js) (link from Tailwind's [docs/configuration](https://tailwindcss.com/docs/configuration)). https://tailwindcss.com/docs/content-configuration#class-detection-in-depth mentions regular expressions... but stops short of actually giving you the actual regexes... or pointing you to where they might be found.

My simple test `views/layout.haml`:
```
!!! 5
%html
  %head
  %body

    - if flash['notice']
      .alert.alert-success
        %p= flash['notice']

    - if flash['error']
      .alert.alert-danger
        %p= flash['error']

    != yield
```

Here's a sequence of the state of things:

- I start out with the HAML as above, and start the watcher

`grep -E '(container|text-2xl|text-3xl)' src/output.css` gives no results

- `%body` -> `%body.container.text-2xl`:

`.container` and `.text-2xl` show up in the output css

- `%body.container.text-2xl` -> `%body.container.text-3xl`:

`.text-3xl` shows up in the output css (which is fine), *but `.text-2xl` remains and does not disappear*

- return to `%body` with no classes

all 3 classes that I previously added remain in the output css: `.container`, `.text-2xl`, `.text-3xl`

- restart the watcher

all 3 classes disappear from the output css

**Comment by wongjn** [comment]
CSS rules remaining in the output CSS after their corresponding classes are removed is expected with `--watch`. I believe there is some assumptions that `--watch` implies that you are developing and that it is quicker to not have to track with class names should be deleted. [This function](https://github.com/tailwindlabs/tailwindcss/blob/master/src/lib/defaultExtractor.js#L23) might give you a clue about the regular expressions in use.

**Reply by jf** [reply]
got it, thank you! That sort of makes sense re not deleting class names... but then perhaps this should be noted down somewhere. And/or the message from the watcher say something besides `Rebuilding...` which would imply a "start all over from scratch" sort of thing.

Thank you for the pointer to the function! It really looks like this is it.
